codeforgreenville/c4g-events
# Limit `/check_api` Invocations to Once Every 15 Minutes per Server

Fixes #5

# Summary
This change only allows for the `/check_api` command to be run on a server/workspace every 15 minutes. If a user from a workspace that has run this command within 15 minutes runs it again, they'll be greeted with this message:
> "This command has been run recently and is on a cooldown period. Please try again in a little while!"

Cooldowns are specific to each workspace, so users from one server aren't subject to a cooldown being enforced against another.

I've also extracted out the instantiation of the SQLite connection(s) from bot.py to database.py to make testing database-dependent functionality a little bit easier.

# Screenshots
A record in the `cooldowns` table where each server (or anything whose ID we'd want to put here) has a timestamp denoting when someone from there can run `/check_api` again.
![Screenshot_2023-10-01_20-54-18](https://github.com/hackgvl/slack-events-bot/assets/44626690/52f33d0a-d23d-4890-b062-f797d0b829b7)


What end users see if they hit the rate-limit.
![Screenshot_2023-10-01_20-43-55](https://github.com/hackgvl/slack-events-bot/assets/44626690/034e29ce-c696-4f85-a5ea-5c21561b7fb7)


# Test Plan
1. Fire up the bot locally and run the `/check_api` command in a workspace that your bot is integrated with.
2. Check for a new record in the `cooldowns` table.
   1. The `accessor` field will be the subdomain related to your workspace (`<accessor>.slack.com`).
   2. The `resource` field will be `check_api`.
   3. The `expires_at` time will be a timestamp in ISO8601 that should be the time 15 minutes from whenever you executed the `/check_api` command (in UTC).
3. Try running the command again. You shoud receive a message about how the command is on cooldown.
4. Adjust the `expires_at` timestamp to be an hour behind what it's currently set to and commit your change. Then try running the command again (it should work).
5. Extra spicy: If you have access to another workspace, verify that the timers are enforced on each server independently (Server B doesn't have a cooldown but Server A does).

I think it would also be good to test the other end-user-facing functionality (like `/add_channel` and `/remove_channel` just to ensure the DB refactor didn't booger anything up.
